internal/bundle/utils.go

```diff
@@ -5,18 +5,97 @@
 package bundle
 
 import (
+	"bytes"
 	"context"
+	"encoding/base64"
+	"encoding/json"
 	"fmt"
 	"io"
 	"os"
 	"path/filepath"
+	"strings"
 
 	"github.com/open-policy-agent/opa/ast"
 	"github.com/open-policy-agent/opa/bundle"
 	"github.com/open-policy-agent/opa/resolver/wasm"
 	"github.com/open-policy-agent/opa/storage"
 )
 
+// PackageFileName is the name of files used to store gzip-compressed bundle data and associated metadata. It is
+// serialized as a JSON object so has a .json extension.
+const PackageFileName = "bundlePackage.json"
+
+// SaveOptions is a list of options which can be set when writing bundle data
+// to disk. Currently, only setting of the bundle's ETag is supported.
+type SaveOptions struct {
+	Etag string
+}
+
+// LoadOptions is a list of options which can be set when loading a bundle from disk.
+// Currently, only setting VerificationConfig is supported.
+type LoadOptions struct {
+	VerificationConfig *bundle.VerificationConfig
+}
+
+// bundlePackage represents a bundle and associated metadata. It implments io.Reader
+// so that it can be serialized to disk easily.
+type bundlePackage struct {
+	Etag   string `json:"etag"`
+	Bundle []byte `json:"bundle"`
+
+	RawBundle io.Reader
+
+	multiReader io.Reader
+}
+
+func (bp *bundlePackage) Read(p []byte) (n int, err error) {
+	if bp.RawBundle == nil {
+		return 0, fmt.Errorf("RawBundle must be set")
+	}
+	if bp.multiReader == nil {
+		b64Reader := &streamingBase64Encoder{
+			input: bp.RawBundle,
+		}
+
+		bp.multiReader = io.MultiReader(
+			strings.NewReader(fmt.Sprintf(`{"etag":"%s","bundle":"`, bp.Etag)),
+			b64Reader,
+			strings.NewReader(`"}`),
+		)
+	}
+
+	return bp.multiReader.Read(p)
+}
+
+// streamingBase64Encoder implements io.Reader and incrementally encodes the
+// data in the input io.Reader as base64.
+type streamingBase64Encoder struct {
+	input io.Reader
+}
+
+func (s *streamingBase64Encoder) Read(p []byte) (n int, err error) {
```

Comment:
This is new and supports streaming the bundle to disk but as base64 encoded bytes. This is needed to load the bundle as []byte from JSON.